lib/src/config_provider/config.dart

```diff
@@ -215,363 +219,689 @@ class Config {
     }
   }
 
-  /// Validates Yaml according to given specs.
-  bool _checkConfigs(YamlMap map, Map<List<String>, Specification> specs) {
-    var result = true;
-    for (final key in specs.keys) {
-      final spec = specs[key];
-      if (checkKeyInYaml(key, map)) {
-        result = result && spec!.validator(key, getKeyValueFromYaml(key, map));
-      } else if (spec!.requirement == Requirement.yes) {
-        _logger.severe("Key '$key' is required.");
-        result = false;
-      } else if (spec.requirement == Requirement.prefer) {
-        _logger.warning("Prefer adding Key '$key' to your config.");
-      }
+  Schema _getRootSchema() {
+    return FixedMapSchema(
+      keys: [
+        FixedMapKey(
+          key: strings.llvmPath,
+          valueSchema: ListSchema<String>(
+            childSchema: StringSchema(),
+            transform: (node) => llvmPathExtractor(node.value),
+          ),
+          defaultValue: (node) => findDylibAtDefaultLocations(),
+          resultOrDefault: (node) => _libclangDylib = node.value as String,
+        ),
+        FixedMapKey(
+            key: strings.output,
+            required: true,
+            valueSchema: OneOfSchema(
+              childSchemas: [
+                _filePathStringSchema(),
+                _outputFullSchema(),
+              ],
+              transform: (node) =>
+                  outputExtractor(node.value, filename, packageConfig),
+              result: (node) {
+                _output = (node.value as OutputConfig).output;
+                _symbolFile = (node.value as OutputConfig).symbolFile;
+              },
+            )),
+        FixedMapKey(
+          key: strings.language,
+          valueSchema: EnumSchema(
+            allowedValues: {strings.langC, strings.langObjC},
+            transform: (node) {
+              if ((node.value == strings.langObjC)) {
+                _logger.severe(
+                    'Objective C support is EXPERIMENTAL. The API may change '
+                    'in a breaking way without notice.');
+                return Language.objc;
+              } else {
+                return Language.c;
+              }
+            },
+          ),
+          defaultValue: (node) => Language.c,
+          resultOrDefault: (node) => _language = node.value as Language,
+        ),
+        FixedMapKey(
+            key: strings.headers,
+            required: true,
+            valueSchema: FixedMapSchema<List<String>>(
+              keys: [
+                FixedMapKey(
+                  key: strings.entryPoints,
+                  valueSchema: ListSchema<String>(childSchema: StringSchema()),
```

Comment:
`ListSchema<String>(childSchema: StringSchema())` String is mentioned twice here.

Can we infer the type argument of `ListSchema` from the `childSchema`?

Reply:
We can't, since child schema might have a `transform` closure, which could change the return value from `extractNode` to  anything (`dynamic`).

The `Generic` is only used for casting the value returned by each of the child objects. In case of ListSchema, it returns a list `<E>[]`. For FixedMapSchema, it returns a map `<dynamic, E>{}`.

Reply:
Should it have two type arguments then? One for the thing being read, and one for the value after it's transformed.

It's rather common to have a list of valid values that is some kind of enum class.

```dart
          BuildMode.fromString(
            config.string(
              BuildMode.configKey,
              validValues: BuildMode.values.map((e) => '$e'),
            ),
          );
```

https://github.com/dart-lang/native/blob/main/pkgs/native_assets_cli/lib/src/model/build_config.dart

That would then be a `ConfigSpecification` with `<String, BuildMode>`. E.g. `String` is the type in JSON/YAML and `BuildMode` is the type in Dart once parsed/transformed.

Reply:
> Should it have two type arguments then? One for the thing being read, and one for the value after it's transformed.

Sure

> It's rather common to have a list of valid values that is some kind of enum class.

We can simply nest EnumSchema inside ListSchema -
```dart
ListSchema(childSchema: EnumSchema(allowedValues: ["a", "b", "c"]))
```

Reply:
> Should it have two type arguments then? One for the thing being read, and one for the value after it's transformed.

Done. The first generic casts the value received before calling transform, and the second parameter casts the value after transform or if transform is not available, casts the same result itself.
The second generic basically decides the value returned by _extractNode.

So if no transform is specified, both the Generic types must be equal